api/src/test/java/org/openmrs/ObsTest.java

```diff
@@ -554,4 +646,272 @@ public void setFormField_shouldRejectAPathContainingTheSeparator() throws Except
+		assertFalse(obs.isDirty());
+		BeanUtils.copyProperties(obs, BeanUtils.cloneBean(obs));
+		assertFalse(obs.isDirty());
+	}
+	
+	/**
+	 * @see Obs#isDirty()
+	 * @verifies return true when any immutable field has been changed
+	 */
+	@Test
+	public void isDirty_shouldReturnTrueWhenAnyImmutableFieldHasBeenChanged() throws Exception {
+		//Set all fields to some random values via reflection
+		List<Field> fields = Reflect.getAllFields(Obs.class);
+		Obs obs = createObs();
+		assertFalse(obs.isDirty());
+		final Integer originalPersonId = obs.getPersonId();
+		//call each setter and check that dirty has been set to true for each
+		for (Field field : fields) {
+			String fieldName = field.getName();
+			if (IGNORED_FIELDS.contains(fieldName)) {
+				continue;
+			}
+
+			if ("personId".equals(fieldName)) {
+				//call setPersonId because it is protected so BeanUtils.setProperty won't work
+				obs.setPersonId((Integer) generateValue(field, true));
+			} else {
+				BeanUtils.setProperty(obs, fieldName, generateValue(field, true));
+			}
+			
+			assertTrue("Obs was not marked as dirty after changing: " + fieldName, obs.isDirty());
+			if ("person".equals(fieldName)) {
+				//Because setPerson updates the personId we need to reset personId to its original value 
+				//that matches that of person otherwise the test will fail for the personId field
+				obs.setPersonId(originalPersonId);
+			}
+			
+			//reset for next field
+			resetObs(obs);
+			
+			assertFalse("Failed to rest Obs.dirty after testing field: " + fieldName, obs.isDirty());
+		}
+	}
+	
+	/**
+	 * @see Obs#isDirty()
+	 * @verifies return false when only mutable fields are changed
+	 */
+	@Test
+	public void isDirty_shouldReturnFalseWhenOnlyMutableFieldsAreChanged() throws Exception {
+		Obs obs = new Obs();
+		obs.setVoided(true);
+		obs.setVoidedBy(new User(1000));
+		obs.setVoidReason("some other reason");
+		obs.setDateVoided(new Date());
+		assertFalse(obs.isDirty());
+	}
+	
+	/**
+	 * @see Obs#isDirty()
+	 * @verifies return true when a field is changed from a non null to a null value
+	 */
+	@Test
+	public void isDirty_shouldReturnTrueWhenAnImmutableFieldIsChangedFromANonNullToANullValue() throws Exception {
+		Obs obs = createObs();
+		assertNotNull(obs.getComment());
+		obs.setComment(null);
+		assertTrue(obs.isDirty());
+	}
+	
+	/**
+	 * @see Obs#isDirty()
+	 * @verifies return true when a field is changed from a null to a non null value
+	 */
+	@Test
+	public void isDirty_shouldReturnTrueWhenAnImmutableFieldIsChangedFromANullToANonNullValue() throws Exception {
+		Obs obs = new Obs();
+		assertNull(obs.getComment());
+		obs.setComment("some non null value");
+		assertTrue(obs.isDirty());
+	}
+	
+	/**
+	 * @see Obs#setFormField(String,String)
+	 * @verifies not mark the obs as dirty when the value has not been changed
+	 */
+	@Test
+	public void setFormField_shouldNotMarkTheObsAsDirtyWhenTheValueHasNotBeenChanged() throws Exception {
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Method names should not contain underscores](https://www.codacy.com/app/openmrs/openmrs-core/file/2318734355/issues/source?bid=3204099&fileBranchId=3204127#l749)